linker/Mono.Linker/Dependencies.cs

```diff
@@ -0,0 +1,274 @@
+				return;
+
+			KeyValuePair<object, object> pair = new KeyValuePair<object, object>(dependency_stack.Count > 0 ? dependency_stack.Peek() : null, o);
+
+			if (!ShouldRecordDependencyInformation(pair.Key) & !ShouldRecordDependencyInformation(pair.Value))
+				return;
+
+			if (pair.Key != pair.Value)
+			{
+				WriteEdge(pair.Key, pair.Value);
+			}
+		}
+
+		public void Push(TypeDefinition definition)
+		{
+			if (writer == null)
+				return;
+
+			if (!WillAssemblyBeModified(definition.Module.Assembly))
+				return;
+
+			PushPrivate(definition);
+		}
+
+		public void Push(AssemblyDefinition definition)
+		{
+			if (writer == null)
+				return;
+
+			if (!WillAssemblyBeModified(definition))
+				return;
+
+			PushPrivate(definition);
+		}
+
+		public void Push(IMemberDefinition definition)
+		{
+			if (writer == null)
+				return;
+
+			if (!WillAssemblyBeModified(definition.DeclaringType.Module.Assembly))
+				return;
+
+			PushPrivate(definition);
+		}
+
+		public void Push(ICustomAttributeProvider provider)
+		{
+			if (writer == null)
+				return;
+
+			// TODO by Mike : Not sure what to do about this one
+			//if (!WillAssemblyBeModified(definition.DeclaringType.Module.Assembly))
+			//	return;
+
+			PushPrivate(provider);
+		}
+
+		public void Push(ExportedType exportedType)
+		{
+			if (writer == null)
+				return;
+
+			// TODO by Mike : Not sure what to do about this one
+			//if (!WillAssemblyBeModified(exportedType.DeclaringType.Module.Assembly))
+			//	return;
+
+			PushPrivate(exportedType);
+		}
+
+		public void Push(IStep step)
+		{
+			if (writer == null)
+				return;
+
+			PushPrivate(step);
+		}
+
+		public void Push(object o)
+		{
+			if (writer == null)
+				return;
+
+			PushPrivate(o);
+		}
+
+		public void Pop()
+		{
+			if (writer == null)
+				return;
+
+			dependency_stack.Pop();
+		}
+
+		public void SaveDependencies()
+		{
+			if (writer == null)
+				return;
+
+			writer.WriteEndElement();
+			writer.WriteEndDocument();
+			writer.Flush();
+			writer.Dispose();
+			zipStream.Dispose();
+			writer = null;
+			zipStream = null;
+			dependency_stack = null;
+		}
+
+		void PushPrivate(object o)
+		{
+			if (dependency_stack.Count > 0)
+				AddDependency(o);
+			dependency_stack.Push(o);
+		}
+
+		void WriteEdge(object b, object e)
+		{
+			writer.WriteStartElement("edge");
+			writer.WriteAttributeString("b", TokenString(b));
+			writer.WriteAttributeString("e", TokenString(e));
+			writer.WriteEndElement();
+		}
+
+		string TokenString(object o)
+		{
+			if (o == null)
+				return "N:null";
+
+			if (o is IMetadataTokenProvider)
+				return (o as IMetadataTokenProvider).MetadataToken.TokenType + ":" + o;
+
+			return "Other:" + o;
+		}
+
+		bool WillAssemblyBeModified(AssemblyDefinition assembly)
+		{
+			// TODO by Mike : Should other values be considered true?
+			return _context.Annotations.GetAction(assembly) == AssemblyAction.Link;
+		}
+
+		bool ShouldRecordDependencyInformation(object o)
```

Comment:
This was my first prototype approach.

I didn't like all of the casting so I switched to the overloads to Push & Begin approach so the type is retained long enough to do whatever smart filtering we want.